### ISSUE TITLE
Exception: Incompatible ssh server (no acceptable macs)

### ISSUE THREAD

**Issue by Sandman3286** [ktbyers/netmiko#2580]
I see a similar issue post for "Exception: Incompatible ssh peer (no acceptable host key) #2556" but this exception is subtly different.  I am trying to access a Juniper device with netmiko (version 3.4.0 and paramiko version 2.8.0) and receive the Traceback below.  Below the Traceback is code used to test and log content.  This only seems to be an issue with Juniper device.  Other devices connect without issue.  Any assistance would be greatly appreciated.  Thanks in advance.

```
Traceback (most recent call last):
  File "netmiko_jun_test.py", line 17, in <module>
    router_connection = ConnectHandler(**juniper_mse) 
  File "/usr/local/lib/python3.6/site-packages/netmiko/ssh_dispatcher.py", line 337, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/netmiko/juniper/juniper.py", line 20, in __init__
    return super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/netmiko/base_connection.py", line 350, in __init__
    self._open()
  File "/usr/local/lib/python3.6/site-packages/netmiko/base_connection.py", line 355, in _open
    self.establish_connection()
  File "/usr/local/lib/python3.6/site-packages/netmiko/base_connection.py", line 935, in establish_connection
    self.remote_conn_pre.connect(**ssh_connect_params)
  File "/usr/local/lib/python3.6/site-packages/paramiko/client.py", line 406, in connect
    t.start_client(timeout=timeout)
  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 660, in start_client
    raise e
  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2083, in run
    self._handler_table[ptype](self, m)
  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2198, in _negotiate_keys
    self._parse_kex_init(m)
  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2438, in _parse_kex_init
    raise SSHException("Incompatible ssh server (no acceptable macs)")
```

Test code:
```
from netmiko import ConnectHandler
import logging

logging.basicConfig(filename='test.log', level=logging.DEBUG)

juniper_device = {
    "device_type": "juniper",
    "host": "device_ip",
    "username": "user_id",
    "password": "user_pwd"
}

router_connection = None
router_connection = ConnectHandler(**juniper_device)
if router_connection:
    router_connection.disconnect()
```

Log content:
```
DEBUG:paramiko.transport:Kex agreed: curve25519-sha256@libssh.org
DEBUG:paramiko.transport:HostKey agreed: ssh-ed25519
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
ERROR:paramiko.transport:Exception: Incompatible ssh server (no acceptable macs)
ERROR:paramiko.transport:Traceback (most recent call last):
ERROR:paramiko.transport:  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2083, in run
ERROR:paramiko.transport:    self._handler_table[ptype](self, m)
ERROR:paramiko.transport:  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2198, in _negotiate_keys
ERROR:paramiko.transport:    self._parse_kex_init(m)
ERROR:paramiko.transport:  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2438, in _parse_kex_init
ERROR:paramiko.transport:    raise SSHException("Incompatible ssh server (no acceptable macs)")
ERROR:paramiko.transport:paramiko.ssh_exception.SSHException: Incompatible ssh server (no acceptable macs)
ERROR:paramiko.transport:
```

**Comment by ktbyers** [comment]
Is there anything above that in the log file?

**Comment by Sandman3286** [comment]
Yes, there is content above that point.  Sorry for the omission.  Here is the entire log:

```
DEBUG:paramiko.transport:starting thread (client mode): 0x2c3daac8
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_2.8.0
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.5
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.5)
DEBUG:paramiko.transport:kex algos:['curve25519-sha256@libssh.org', 'ecdh-sha2-nistp521', 'ecdh-sha2-nistp384', 'ecdh-sha2-nistp256'] server key:['ssh-rsa', 'rsa-sha2-512', 'rsa-sha2-256', 'ecdsa-sha2-nistp256', 'ssh-ed25519'] client encrypt:['chacha20-poly1305@openssh.com', 'aes256-gcm@openssh.com', 'aes128-gcm@openssh.com', 'aes256-ctr', 'aes192-ctr', 'aes128-ctr'] server encrypt:['chacha20-poly1305@openssh.com', 'aes256-gcm@openssh.com', 'aes128-gcm@openssh.com', 'aes256-ctr', 'aes192-ctr', 'aes128-ctr'] client mac:['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com'] server mac:['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com'] client compress:['none', 'zlib@openssh.com'] server compress:['none', 'zlib@openssh.com'] client lang:[''] server lang:[''] kex follows?False
DEBUG:paramiko.transport:Kex agreed: curve25519-sha256@libssh.org
DEBUG:paramiko.transport:HostKey agreed: ssh-ed25519
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
ERROR:paramiko.transport:Exception: Incompatible ssh server (no acceptable macs)
ERROR:paramiko.transport:Traceback (most recent call last):
ERROR:paramiko.transport:  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2083, in run
ERROR:paramiko.transport:    self._handler_table[ptype](self, m)
ERROR:paramiko.transport:  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2198, in _negotiate_keys
ERROR:paramiko.transport:    self._parse_kex_init(m)
ERROR:paramiko.transport:  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2438, in _parse_kex_init
ERROR:paramiko.transport:    raise SSHException("Incompatible ssh server (no acceptable macs)")
ERROR:paramiko.transport:paramiko.ssh_exception.SSHException: Incompatible ssh server (no acceptable macs)
ERROR:paramiko.transport:
```

**Comment by ktbyers** [comment]
Can you edit your Paramiko transport.py file (temporarily) to add the following print statements. This should be around line 2438 in this transport.py file.

/usr/local/lib/python3.6/site-packages/paramiko/transport.py

```python
        print(client_mac_algo_list)    # add
        print(self.preferred_macs)    # add
        print(server_mac_algo_list)  # add
        if (len(agreed_local_macs) == 0) or (len(agreed_remote_macs) == 0):
            raise SSHException("Incompatible ssh server (no acceptable macs)")
```

**Comment by ktbyers** [comment]
To make it easier to read here are the macs in the output above:

client mac:['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com'] 

server mac:['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com']

**Comment by Sandman3286** [comment]
Output from the print statements:
```
['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com']
('hmac-sha1',)
['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com']
```

Preferred list does not match anything in the client or server mac list.  :(

The interesting this is, creating a virtual environment on the same server via :
```
python3 -m venv venv
source venv/bin/activate
pip install netmiko
```

Works, and the output of the print statements from that same paramiko file (transport.py) in the venv produces:
```
['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com']
('hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1', 'hmac-md5', 'hmac-sha1-96', 'hmac-md5-96')
['hmac-sha2-512-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-512', 'hmac-sha2-256', 'umac-128@openssh.com']
```

Very different preferred list.

**Comment by ktbyers** [comment]
Okay, what version of the `cryptography` library are on the two different python environments (the one that was failing and the new venv)?

**Comment by Sandman3286** [comment]
To answer, both environments are running cryptography 35.0.0.  But we've discovered the root cause of our issue.  We found that a custom driver is doing:

`paramiko.Transport._preferred_macs = ("hmac-sha1",)`

And once netmiko is imported that gets set in our running environment.  Anything that not have hmac-sha1 in their list fails to connect.  In this case it is Juniper devices for us.

Thank you very much for your prompt responses to our post.
